### ISSUE TITLE
execute_script does not like objects returned by _element decorator

### ISSUE THREAD

**Issue by tk8817** [cheezy/page-object#400]
## Issue 

When invoking `@browser.execute_script`, we have found that only a 'pure' Watir element definition will work. Using an element defined by `page-object` and retrieved via  `_element` will result in a strange error. 

#### Define an element with `page-object` and retrieve it using `_element`. This looks good!
```ruby
select_list( :gender,          id: 'gender') # in page object

[8] pry(#<Newsletter>)> gender_element
=> #<Watir::Select:0x..fc201075737c4d7a8 located=false selector={:id=>"gender", :tag_name=>"select"}>
```

#### Define an element with `watir` directly. This looks good!
```ruby
[9] pry(#<Newsletter>)> watir_gender_element = browser.select_list(id: 'gender')
=> #<Watir::Select:0x..f8ec8553fd8c18adc located=false selector={:id=>"gender", :tag_name=>"select"}>
```

 #### Try use the `watir` element in tandem with `execute_script`. This looks good!
```ruby
 pry(#<Newsletter>)> @browser.execute_script("$(arguments[0]).val('Male').change();", watir_gender_element)
=> nil
```

#### Try use the `page-object` element in tandem with `execute_script`. Oh no, error!
```ruby
pry(#<Newsletter>)> browser.execute_script("$(arguments[0]).val('Male').change();", gender_element)
    Selenium::WebDriver::Error::UnknownError: unknown error: Syntax error, unrecognized expression: #<PageObject::Elements::SelectList:0x007f930e272bf0>
    (Session info: chrome=54.0.2840.98)
(Driver info: chromedriver=2.25.426935 (820a95b0b81d33e42712f9198c215f703412e1a1),platform=Mac OS X 10.12.1 x86_64)
from /Users/fbarann/.gem/ruby/2.3.1/gems/selenium-webdriver-3.0.1/lib/selenium/webdriver/remote/response.rb:69:in `assert_ok'
```

## Investigation 

We did some digging in pry and after following a twisting trail of calls to different `execute_script` 
the error is raised by the following bit of code in <Selenium::WebDriver::Chrome::Bridge>. 

```ruby
def raw_execute(command, opts = {}, command_hash = nil)
          verb, path = COMMANDS[command] || raise(ArgumentError, "unknown command: #{command.inspect}")
          path = path.dup

          path[':session_id'] = @session_id if path.include?(':session_id')

          begin
            opts.each { |key, value| path[key.inspect] = escaper.escape(value.to_s) }
          rescue IndexError
            raise ArgumentError, "#{opts.inspect} invalid for #{command.inspect}"
          end

          puts "-> #{verb.to_s.upcase} #{path}" if $DEBUG
          http.call verb, path, command_hash
        end
```

#### Pry'ing around that method with the `page-object` element
```ruby
[5] pry(#<Newsletter>)> browser.execute_script("$(arguments[0]).val('Male').change();", gender_element)
From: (pry) @ line 4 Selenium::WebDriver::Remote::Bridge#raw_execute:
     3: def raw_execute(command, opts = {}, command_hash = nil)
 =>  4:   binding.pry
     5:   verb, path = COMMANDS[command] || raise(ArgumentError, "unknown command: #{command.inspect}")
     6:   path = path.dup
     7:
     8:   path[':session_id'] = @session_id if path.include?(':session_id')
     9:
    10:     begin
    11:       opts.each { |key, value| path[key.inspect] = escaper.escape(value.to_s) }
    12:     rescue IndexError
    13:       raise ArgumentError, "#{opts.inspect} invalid for #{command.inspect}"
    14:     end
    15:
    16:     puts "-> #{verb.to_s.upcase} #{path}" if $DEBUG
    17:     http.call verb, path, command_hash
    18:   end
[1] pry(#<Selenium::WebDriver::Chrome::Bridge>)> command
=> :executeScript
[2] pry(#<Selenium::WebDriver::Chrome::Bridge>)> command_hash
=> {:script=>"$(arguments[0]).val('Male').change();",
 :args=>
  [#<Watir::Select:0x..f8df3d36b8bd2b0a2 located=false selector={:id=>"gender", :tag_name=>"select"}>]}
[3] pry(#<Selenium::WebDriver::Chrome::Bridge>)> verb, path = COMMANDS[command] || raise(ArgumentError, "unknown command: #{command.inspect}")
=> [:post, "session/:session_id/execute"]
[4] pry(#<Selenium::WebDriver::Chrome::Bridge>)> verb
=> :post
[5] pry(#<Selenium::WebDriver::Chrome::Bridge>)> path
=> "session/:session_id/execute"
[6] pry(#<Selenium::WebDriver::Chrome::Bridge>)> path = path.dup
=> "session/:session_id/execute"
[7] pry(#<Selenium::WebDriver::Chrome::Bridge>)> path[':session_id'] = @session_id if path.include?(':session_id')
=> "cb462a7b108758601beba69a80e81da1"
[8] pry(#<Selenium::WebDriver::Chrome::Bridge>)> opts.each { |key, value| path[key.inspect] = escaper.escape(value.to_s) }
=> {}
[9] pry(#<Selenium::WebDriver::Chrome::Bridge>)> begin
[9] pry(#<Selenium::WebDriver::Chrome::Bridge>)*   opts.each { |key, value| path[key.inspect] = escaper.escape(value.to_s) }
[9] pry(#<Selenium::WebDriver::Chrome::Bridge>)* rescue IndexError
[9] pry(#<Selenium::WebDriver::Chrome::Bridge>)*   raise ArgumentError, "#{opts.inspect} invalid for #{command.inspect}"
[9] pry(#<Selenium::WebDriver::Chrome::Bridge>)* end
=> {}
[10] pry(#<Selenium::WebDriver::Chrome::Bridge>)> puts "-> #{verb.to_s.upcase} #{path}" if $DEBUG
=> nil
[11] pry(#<Selenium::WebDriver::Chrome::Bridge>)> http.call verb, path, command_hash
Selenium::WebDriver::Error::UnknownError: unknown error: Syntax error, unrecognized expression: #<PageObject::Elements::SelectList:0x007fb5b12bdb30>
  (Session info: chrome=54.0.2840.98)
  (Driver info: chromedriver=2.25.426935 (820a95b0b81d33e42712f9198c215f703412e1a1),platform=Mac OS X 10.12.1 x86_64)
from /Users/fbarann/.gem/ruby/2.3.1/gems/selenium-webdriver-3.0.1/lib/selenium/webdriver/remote/response.rb:69:in `assert_ok'
```

#### Pry'ing around that method with the `watir` element
```ruby
[5] pry(#<Newsletter>)> browser.execute_script("$(arguments[0]).val('Male').change();", @browser.select_list(id: 'gender'))
From: (pry) @ line 4 Selenium::WebDriver::Remote::Bridge#raw_execute:
     3: def raw_execute(command, opts = {}, command_hash = nil)
 =>  4:   binding.pry
     5:   verb, path = COMMANDS[command] || raise(ArgumentError, "unknown command: #{command.inspect}")
     6:   path = path.dup
     7:
     8:   path[':session_id'] = @session_id if path.include?(':session_id')
     9:
    10:     begin
    11:       opts.each { |key, value| path[key.inspect] = escaper.escape(value.to_s) }
    12:     rescue IndexError
    13:       raise ArgumentError, "#{opts.inspect} invalid for #{command.inspect}"
    14:     end
    15:
    16:     puts "-> #{verb.to_s.upcase} #{path}" if $DEBUG
    17:     http.call verb, path, command_hash
    18:   end
[1] pry(#<Selenium::WebDriver::Chrome::Bridge>)> verb, path = COMMANDS[command] || raise(ArgumentError, "unknown command: #{command.inspect}")
=> [:get, "session/:session_id/window_handle"]
[2] pry(#<Selenium::WebDriver::Chrome::Bridge>)> path = path.dup
=> "session/:session_id/window_handle"
[3] pry(#<Selenium::WebDriver::Chrome::Bridge>)> path[':session_id'] = @session_id if path.include?(':session_id')
=> "cb462a7b108758601beba69a80e81da1"
[4] pry(#<Selenium::WebDriver::Chrome::Bridge>)> begin
[4] pry(#<Selenium::WebDriver::Chrome::Bridge>)*   opts.each { |key, value| path[key.inspect] = escaper.escape(value.to_s) }
[4] pry(#<Selenium::WebDriver::Chrome::Bridge>)* rescue IndexError
[4] pry(#<Selenium::WebDriver::Chrome::Bridge>)*   raise ArgumentError, "#{opts.inspect} invalid for #{command.inspect}"
[4] pry(#<Selenium::WebDriver::Chrome::Bridge>)* end
=> {}
[5] pry(#<Selenium::WebDriver::Chrome::Bridge>)> puts "-> #{verb.to_s.upcase} #{path}" if $DEBUG
=> nil
[6] pry(#<Selenium::WebDriver::Chrome::Bridge>)> http.call verb, path, command_hash
=> #<Selenium::WebDriver::Remote::Response:0x007fb5b1f16328 @code=200, @payload={"sessionId"=>"cb462a7b108758601beba69a80e81da1", "status"=>0, "value"=>"CDwindow-4f9d8b36-aee7-4063-b69d-00809e907025"}>
```

#### The key difference we can see is the command hash

`page-object`
```ruby
pry(#<Selenium::WebDriver::Chrome::Bridge>)> command_hash
=> {:script=>"$(arguments[0]).val('Male').change();",
 :args=>
  [#<Watir::Select:0x..f8df3d36b8bd2b0a2 located=false selector={:id=>"gender", :tag_name=>"select"}>]}
```

`watir`
```ruby
[6] pry(#<Selenium::WebDriver::Chrome::Bridge>)> command_hash
=> {:script=>"$(arguments[0]).val('Male').change();",
 :args=>[#<Selenium::WebDriver::Element:0x..f80f2e25765c2cdbc id="0.24227458152887338-1">]}
```

**Comment by jkotests** [comment]
Are you saying that your page object is calling `Watir::Browser#execute_script` directly?

Given that you are inside a page object, I would think you should be calling the `PageObject#execute_script`. The PageObject method converts elements in the arguments Array to their native type:

~~~~~~~~
def execute_script(script, *args)
  args.map! { |e| e.kind_of?(PageObject::Elements::Element) ? e.element : e }
  platform.execute_script(script, *args)
end
~~~~~~~~

In other words, your page object should be doing just `execute_script` instead of `@browser.execute_script`.

**Comment by tk8817** [comment]
Using your recommended fix of not calling @browser, we were able to get everything working.  
Why were we using @browser in the first place? We are preparing to leave the page-object ecosystem, are are bypassing as many page-object specific methods as possible. Clearly that caused us some harm here! 

Interestingly, your findings led us to something else. Based on that `args.map!` code, its clear `page-object` allows you to send it a page-obect element and not an underlying driver element. 

If you call 
```ruby
select_list( :gender,          id: 'gender') # in page object

 pry(#<Newsletter>)> @browser.execute_script("$(arguments[0]).val('Male').change();", gender)
=> nil
```

you actually get the method to be executed. However, you do not see the desired result in the browser. seemingly the page object version of the method is actually broken in some way.

**Comment by cheezy** [comment]
If you pass the page-object version of execute_script a page-object Element it simply passes the native watir element along.  Otherwise, it passes whatever you send it.  The method is not broken.

**Comment by tk8817** [comment]
@cheezy but based on your code in the map, the method should work with both `gender` and `gender_element` no? In the end they should both get passed on as the native watir element. 

What is even more strange is that if you use `gender` while you don't get an error, you don't see the desired effect in the browser. When we traced through the `watir` implementation of `execute-script` we see that they call `wd` which actually passes the Selenium WebDriver element on.

**Comment by cheezy** [comment]
Based on the code you provided gender_element is a SelectList element and it will return the underlying watir element.  gender is a method that will return the value of the selected Option in the select list.  They are not equivalent.

**Comment by tk8817** [comment]
apologies - we are only barking up the `gender` tree based on a misconception above that `page-object` might actually want its own element type sent across. Everything we were doing in the investigation & up to @jkotests response was with `gender_element` , which seemingly is the correct approach. 

As such, the initial issue stands. You will get an error when trying to call `page-object` variant of `execute_script` with gender_element:

```pry(#<Newsletter>)> browser.execute_script("$(arguments[0]).val('Male').change();", gender_element)
    Selenium::WebDriver::Error::UnknownError: unknown error: Syntax error, unrecognized expression: #<PageObject::Elements::SelectList:0x007f930e272bf0>
    (Session info: chrome=54.0.2840.98)
(Driver info: chromedriver=2.25.426935 (820a95b0b81d33e42712f9198c215f703412e1a1),platform=Mac OS X 10.12.1 x86_64)
from /Users/fbarann/.gem/ruby/2.3.1/gems/selenium-webdriver-3.0.1/lib/selenium/webdriver/remote/response.rb:69:in `assert_ok'```

**Comment by cheezy** [comment]
Do you have an example you can provide of this not working?  I have personally used this many times and have never had a problem with it.

**Comment by tk8817** [comment]
This is no longer happening in the latest versions and I am not sure I was articulating our issue well. This can be safely closed.
